### PR TITLE
Adding PickupAPI

### PR DESCRIPTION
**Pull request by louis1706**
- This branch is from Exiled::DevBranch (it's can re add stuff we don't want in this new version it's need double check for every line changed)
- Some file can be empty (this means they will need to be deleted or re added if the functionality is missing)

### CODE REVIEW

File: Exiled.Events/Patches/Fixes/GrenadePropertiesFix.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            // {
+            //     Log.Error("Item is not Throwable, should never happen");
+            //     return;
+            // }
+            // Projectile projectile = throwable.Projectile;
+            // ThrownProjectile baseProjectile = projectile.Base;
+            // baseProjectile.transform.position = this.Owner.PlayerCameraReference.position;
+            // baseProjectile.transform.rotation = this.Owner.PlayerCameraReference.rotation;
+            // baseProjectile.gameObject.SetActive(true);
+            // projectile.Spawned = true;
```
Comment (iopietro) [inline Exiled.Events/Patches/Fixes/GrenadePropertiesFix.cs:59]: Divide this code in the instructions below, above their corresponding IL.

File: Exiled.Events/Patches/Generic/InventoryControlPatch.cs [hunk 2]
```diff
@@ -69,7 +74,16 @@ internal static class InventoryControlAddPatch
...
         }
 
-        private static void AddItem(Player player, ItemBase item) => player?.ItemsValue.Add(Item.Get(item));
+        private static void AddItem(Player player, ItemBase itemBase, ItemPickupBase itemPickupBase)
+        {
+            Item item = Item.Get(itemBase);
+            if (itemPickupBase != null)
+            {
+                item.Scale = itemPickupBase.transform.localScale;
+            }
```
Comment (iopietro) [inline Exiled.Events/Patches/Generic/InventoryControlPatch.cs:83]:
```suggestion
            Item item = Item.Get(itemBase);

            if (itemPickupBase != null)
                item.Scale = itemPickupBase.transform.localScale;
```

File: Exiled.Events/Patches/Generic/PickupControlPatch.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        private static IEnumerable<CodeInstruction> Transpiler(
+            IEnumerable<CodeInstruction> instructions,
+            ILGenerator generator)
+        {
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Shared.Rent(instructions);
+
+            const int offset = -1;
+            int index = newInstructions.FindIndex(i =>
+                i.opcode == OpCodes.Ldarg_3) + offset;
```
Comment (iopietro) [inline Exiled.Events/Patches/Generic/PickupControlPatch.cs:44]:
```suggestion
            int index = newInstructions.FindIndex(i => i.opcode == OpCodes.Ldarg_3) + offset;
```

File: Exiled.Events/Patches/Generic/StaticPickupsControlPatch.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
+        {
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Shared.Rent(instructions);
+
+            int offset = 1;
+            int index = newInstructions.FindLastIndex(i => i.Calls(Method(typeof(Transform), nameof(Transform.SetParent), new[] { typeof(Transform) }))) + offset;
+
+            newInstructions.InsertRange(index, new[]
+            {
+                new CodeInstruction(OpCodes.Ldloc_1),
```
Comment (iopietro) [inline Exiled.Events/Patches/Generic/StaticPickupsControlPatch.cs:42]: Add some comments to this transpiler.

File: Exiled.API/Features/Items/ExplosiveGrenade.cs [hunk 5]
```diff
@@ -122,14 +150,17 @@ public void SpawnActive(Vector3 position, Player owner = null)
...
+                BurnDuration = BurnDuration,
+                DeafenDuration = DeafenDuration,
+                ConcussDuration = ConcussDuration,
+                FuseTime = FuseTime,
+                PinPullTime = PinPullTime,
+                Repickable = Repickable,
+            };
 
             return cloneableItem;
         }
```
Comment (iopietro) [inline Exiled.API/Features/Items/ExplosiveGrenade.cs:166]: This should be done in one line.

File: Exiled.API/Features/Items/FlashGrenade.cs [hunk 6]
```diff
@@ -95,26 +95,27 @@ public void SpawnActive(Vector3 position, Player owner = null)
...
+            FlashGrenade cloneableItem = new()
+            {
+                BlindCurve = BlindCurve,
+                SurfaceDistanceIntensifier = SurfaceDistanceIntensifier,
+                DeafenCurve = DeafenCurve,
+                FuseTime = FuseTime,
+            };
 
             return cloneableItem;
         }
```
Comment (iopietro) [inline Exiled.API/Features/Items/FlashGrenade.cs:113]: This should be done in one line.

File: Exiled.API/Features/Items/Flashlight.cs [hunk 7]
```diff
@@ -52,21 +52,23 @@ public bool Active
...
         public override Item Clone()
         {
-            Flashlight cloneableItem = new();
-            cloneableItem.Active = Active;
+            Flashlight cloneableItem = new()
+            {
+                Active = Active,
+            };
             return cloneableItem;
         }
```
Comment (iopietro) [inline Exiled.API/Features/Items/Flashlight.cs:66]: This should be done in one line.

File: Exiled.Events/Handlers/Internal/Round.cs [hunk 8]
```diff
@@ -44,8 +38,6 @@ public static void OnRestartingRound()
         {
             MultiAdminFeatures.CallEvent(MultiAdminFeatures.EventType.ROUND_END);
 
-            Scp173Role.TurnedPlayers.Clear();
-            Scp096Role.TurnedPlayers.Clear();
```
Comment (iopietro) [inline Exiled.Events/Handlers/Internal/Round.cs:48]: We don't need to clear those list anymore?

**Review by iopietro — CHANGES_REQUESTED** [review]

File: Exiled.API/Features/Items/Firearm.cs [hunk 9]
```diff
@@ -253,18 +254,14 @@ public void RemoveAttachment(AttachmentIdentifier identifier)
             if (!Attachments.Any(attachment => (attachment.Name == identifier.Name) && attachment.IsEnabled))
                 return;
 
-            uint code = identifier.Code == 0
-                ? AvailableAttachments[Type].FirstOrDefault(
-                    attId =>
-                        attId.Name == identifier.Name).Code
-                : identifier.Code;
+            uint code = identifier.Code;
```
[the remark this answers is not in the document]
Comment (louis1706) [reply Exiled.API/Features/Items/Firearm.cs:257]: https://github.com/louis1706/EXILED/commit/7c6074c8e3e0adf1f5aa7dfe857d6780b1b9441a
Optimise from IRacle

File: Exiled.API/Features/Items/Firearm.cs [hunk 10]
```diff
@@ -278,14 +275,14 @@ public void RemoveAttachment(AttachmentName attachmentName)
             if (firearmAttachment is null)
```
[the remark this answers is not in the document]
Comment (louis1706) [reply Exiled.API/Features/Items/Firearm.cs]: Remove

**Pull request merged**